# Data Type Optimization
With the data we are working in this course, it is easy to take advantage of the abundant computer resources offered. Most of the datasets easily fit within memory. But what happens if your dataset is massive, say 64GB! You can either get a bigger machine, or you can see if optimizing how pandas handles the dataset provides another solution.

Here we will take the `clean_08.csv` dataset we produced in the last lesson, *Fixing Data Types*, and show how altering the data types can shrink the memory footprint of a DataFrame.

In [1]:
import pandas as pd

In [2]:
# read the clean_08 CSV
df = pd.read_csv("clean_08.csv")

In [3]:
# use .info() to view the current Dtypes, and the memory usage.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 987 entries, 0 to 986
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   model                 987 non-null    object 
 1   displ                 987 non-null    float64
 2   cyl                   987 non-null    int64  
 3   trans                 987 non-null    object 
 4   drive                 987 non-null    object 
 5   fuel                  987 non-null    object 
 6   veh_class             987 non-null    object 
 7   air_pollution_score   987 non-null    float64
 8   city_mpg              987 non-null    float64
 9   hwy_mpg               987 non-null    float64
 10  cmb_mpg               987 non-null    float64
 11  greenhouse_gas_score  987 non-null    int64  
 12  smartway              987 non-null    object 
dtypes: float64(5), int64(2), object(6)
memory usage: 100.4+ KB


### Numerical Optimization
Currently the DataFrame is a mixture of objects, float64, and int64 data types. We can also see the memory usage - `100.4+ KB`

Let's see what values are present in the `*_mpg` columns

In [4]:
# find city_mpg value counts
df.city_mpg.value_counts()

17.0    109
16.0     96
20.0     96
15.0     90
19.0     87
18.0     81
13.0     73
14.0     71
21.0     64
12.0     57
22.0     44
23.0     21
24.0     20
11.0     17
10.0     14
25.0      8
26.0      8
27.0      8
9.0       6
29.0      5
8.0       3
34.0      3
28.0      2
40.0      1
35.0      1
33.0      1
48.0      1
Name: city_mpg, dtype: int64

In [5]:
# find hwy_mpg value counts
df.hwy_mpg.value_counts()

24.0    111
26.0     82
25.0     77
28.0     70
19.0     70
20.0     69
27.0     60
22.0     59
18.0     58
29.0     49
23.0     48
17.0     41
31.0     41
30.0     33
21.0     26
16.0     25
33.0     19
32.0     16
15.0     10
14.0      5
35.0      5
34.0      4
36.0      3
13.0      3
45.0      2
37.0      1
Name: hwy_mpg, dtype: int64

In [6]:
# find cmb_mpg value counts
df.cmb_mpg.value_counts()

22.0    101
20.0     94
19.0     89
21.0     84
15.0     83
18.0     77
24.0     67
14.0     64
17.0     61
23.0     59
16.0     53
25.0     36
26.0     24
27.0     17
13.0     16
12.0     15
28.0     13
29.0     12
11.0      7
32.0      4
10.0      3
31.0      3
34.0      2
42.0      1
30.0      1
46.0      1
Name: cmb_mpg, dtype: int64

Even though the DataFrame labels it as float64, when inspecting each one they return `int64`. Let's change it to make it official.

In [7]:
# Change city_mpg, hwy_mpg, cmb_mpg to be an int using .astype()
df["city_mpg"] = df["city_mpg"].astype("int")
df["hwy_mpg"] = df["hwy_mpg"].astype("int")
df["cmb_mpg"] = df["cmb_mpg"].astype("int")

In [8]:
# df info to view data type and memory usage changes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 987 entries, 0 to 986
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   model                 987 non-null    object 
 1   displ                 987 non-null    float64
 2   cyl                   987 non-null    int64  
 3   trans                 987 non-null    object 
 4   drive                 987 non-null    object 
 5   fuel                  987 non-null    object 
 6   veh_class             987 non-null    object 
 7   air_pollution_score   987 non-null    float64
 8   city_mpg              987 non-null    int64  
 9   hwy_mpg               987 non-null    int64  
 10  cmb_mpg               987 non-null    int64  
 11  greenhouse_gas_score  987 non-null    int64  
 12  smartway              987 non-null    object 
dtypes: float64(2), int64(5), object(6)
memory usage: 100.4+ KB


Well that did not change anything. Instead of an `int64`, let's change them to be `int8`. The values for each column only range from 8 - 48. Use `.describe()` to view the min/max of each column

In [9]:
df[["city_mpg", "hwy_mpg", "cmb_mpg"]].describe()

city_mpg     hwy_mpg     cmb_mpg
count  987.000000  987.000000  987.000000
mean    17.386018   24.038501   19.788247
std      4.088018    4.753406    4.251565
min      8.000000   13.000000   10.000000
25%     15.000000   20.000000   17.000000
50%     17.000000   24.000000   20.000000
75%     20.000000   27.000000   22.000000
max     48.000000   45.000000   46.000000

In [10]:
# Change the data type to be an int8
df["city_mpg"] = df["city_mpg"].astype("int8")
df["hwy_mpg"] = df["hwy_mpg"].astype("int8")
df["cmb_mpg"] = df["cmb_mpg"].astype("int8")

In [11]:
# df info to view data type and memory usage changes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 987 entries, 0 to 986
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   model                 987 non-null    object 
 1   displ                 987 non-null    float64
 2   cyl                   987 non-null    int64  
 3   trans                 987 non-null    object 
 4   drive                 987 non-null    object 
 5   fuel                  987 non-null    object 
 6   veh_class             987 non-null    object 
 7   air_pollution_score   987 non-null    float64
 8   city_mpg              987 non-null    int8   
 9   hwy_mpg               987 non-null    int8   
 10  cmb_mpg               987 non-null    int8   
 11  greenhouse_gas_score  987 non-null    int64  
 12  smartway              987 non-null    object 
dtypes: float64(2), int64(2), int8(3), object(6)
memory usage: 80.1+ KB


Now we are getting somewhere! We just changed the memory usage from `100.4+ KB` to `80.1+ KB` by changing how we are storing our int values.

How about changing how we store strings?

### String Optimization
Look at the value counts of each `object` data type: `trans`, `drive`, `fuel`, `veh_class`, `smartway`, and `model`.

In [12]:
# find trans value counts
df["trans"].value_counts()

Auto-L4    176
Auto-S6    162
Auto-L5    157
Man-6      142
Man-5      123
Auto-S5     71
Auto-L6     56
Auto-AV     45
Auto-S4     21
Auto-L7     15
Auto-S7     11
Auto-6       4
S8           4
Name: trans, dtype: int64

In [13]:
# find drive value counts
df["drive"].value_counts()

2WD    662
4WD    325
Name: drive, dtype: int64

In [14]:
# find fuel value counts
df["fuel"].value_counts()

Gasoline    984
CNG           1
ethanol       1
gas           1
Name: fuel, dtype: int64

In [15]:
# find veh_class value counts
df["veh_class"].value_counts()

small car        333
SUV              280
midsize car      138
pickup            83
station wagon     60
large car         55
van               21
minivan           17
Name: veh_class, dtype: int64

In [16]:
# find smartway value counts
df["smartway"].value_counts()

no     607
yes    380
Name: smartway, dtype: int64

In [17]:
# find model value counts
df["model"].value_counts()

NISSAN Altima             12
HONDA Accord              11
FORD Ranger               10
DODGE RAM 1500             9
DODGE Dakota               8
                          ..
MERCEDES-BENZ SL55 AMG     1
MERCEDES-BENZ SL550        1
MERCEDES-BENZ SL600        1
MERCEDES-BENZ SL65 AMG     1
ACURA MDX                  1
Name: model, Length: 377, dtype: int64

Except for `model`, all of the object types have 2 - 13 unique values. In pandas there is a specialized data type called [Categorical](https://pandas.pydata.org/docs/user_guide/categorical.html#). Categorical data types are useful when you have object columns with a low number of unique values. You can create them as categories, and pandas will store those columns more efficiently.

Let's change the 5 `object` columns to `category`

In [18]:
# assign trans, drive, fuel, veh_class, and smartway to "category" using .astype()
df["trans"] = df["trans"].astype("category")
df["drive"] = df["drive"].astype("category")
df["fuel"] = df["fuel"].astype("category")
df["veh_class"] = df["veh_class"].astype("category")
df["smartway"] = df["smartway"].astype("category")

In [19]:
# df info to view data type and memory usage changes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 987 entries, 0 to 986
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype   
---  ------                --------------  -----   
 0   model                 987 non-null    object  
 1   displ                 987 non-null    float64 
 2   cyl                   987 non-null    int64   
 3   trans                 987 non-null    category
 4   drive                 987 non-null    category
 5   fuel                  987 non-null    category
 6   veh_class             987 non-null    category
 7   air_pollution_score   987 non-null    float64 
 8   city_mpg              987 non-null    int8    
 9   hwy_mpg               987 non-null    int8    
 10  cmb_mpg               987 non-null    int8    
 11  greenhouse_gas_score  987 non-null    int64   
 12  smartway              987 non-null    category
dtypes: category(5), float64(2), int64(2), int8(3), object(1)
memory usage: 47.8+ KB


Wow! By changing those columns to categories, we further reduced our dataset from `80.1+ KB` to `47.8+ KB`. We effectively reduced the memory usage by 50%. While it may not be important in smaller datasets such as this one, you can really see the power when working on large data.

For more information, check out pandas [use-efficient-datatypes](https://pandas.pydata.org/docs/user_guide/scale.html#use-efficient-datatypes) section.